File: dart/structural/flyweight/guru_example/bin/flyweight_factory.dart
import 'flyweight.dart';

class FlyweightFactory {
  final Map<String, Flyweight> _flyweights = {};

  FlyweightFactory(List<List<String>> initialFlyweights) {
    for (var state in initialFlyweights) {
      _flyweights[_getKey(state)] = Flyweight(state);
    }
  }

  String _getKey(List<String> state) {
    return state.join('_');
  }

  Flyweight getFlyweight(List<String> sharedState) {
    final key = _getKey(sharedState);

    if (!_flyweights.containsKey(key)) {
      print('FlyweightFactory: $key is not exist.');
      _flyweights[key] = Flyweight(sharedState);
    } else {
      print('FlyweightFactory: $key is already exist.');
    }

    return _flyweights[key];
  }

  void listFlyweights() {
    final count = _flyweights.keys.length;
    print('FlyweightFactory: now have $count flyweights.');
    for (var key in _flyweights.keys) {
      print(key);
    }
  }
}

File: typescript/structural/flyweight/guru_example/flyweightFactory.ts
import {Flyweight} from "./flyweight";

export class FlyweightFactory {
  private flyweights: {[key: string]: Flyweight} = <any>{}

  constructor(initialFlyweights: string[][]) {
    for (const state of initialFlyweights) {
      this.flyweights[this.getKey(state)] = new Flyweight(state)
    }
  }

  private getKey(state: string[]): string {
    return state.join('_')
  }

  public getFlyweight(sharedState: string[]): Flyweight {
    const key = this.getKey(sharedState)

    if (!(key in this.flyweights)) {
      console.log(`FlyweightFactory: ${key} is not find in flyweights`)
      this.flyweights[key] = new Flyweight(sharedState)
    } else {
      console.log(`FlyweightFactory: ${key} already exist.`)
    }

    return this.flyweights[key]
  }

  public listFlyweight(): void {
    const count = Object.keys(this.flyweights).length
    console.log(`FlyweightFactory: now have ${count} flyweight.`)
    for (const key in this.flyweights) {
      console.log(key)
    }
  }
}
